cvmfs/catalog_mgr_client.cc

```diff
@@ -86,179 +88,246 @@ shash::Any ClientCatalogManager::GetRootHash() {
+    result->root_ctlg_location = kCtlgLocationServer;
+    offline_mode_ = false;
 
-    *catalog_hash = cache_hash;
+    return catalog::kLoadNew;
+    }
+  }
+  LogCvmfs(kLogCache, kLogDebug,
+            "failed fetch manifest from server: "
+            "manifest too old or server unreachable (%d - %s)",
+            manifest_failure, manifest::Code2Ascii(manifest_failure));
+
+  if (manifest_failure == manifest::kFailOk
+      && ensemble.manifest->revision() == local_newest_revision) {
+      offline_mode_ = false;
+  } else {
     offline_mode_ = true;
-    return success_code;
   }
+  result->hash = local_newest_hash;
+  result->root_ctlg_revision = local_newest_revision;
+
+  return success_code;
+}
 
-  manifest_ = new manifest::Manifest(*ensemble.manifest);
-
-  offline_mode_ = false;
-  cvmfs_path += " (" + ensemble.manifest->catalog_hash().ToString() + ")";
-  LogCvmfs(kLogCache, kLogDebug, "remote checksum is %s",
-           ensemble.manifest->catalog_hash().ToString().c_str());
-
-  // Short way out, use cached copy
-  if (ensemble.manifest->catalog_hash() == cache_hash) {
-    LoadError success_code = catalog::kLoadUp2Date;
-
-    // Has the breadcrumb been updated externally?
-    std::map<PathString, shash::Any>::const_iterator iter =
-      mounted_catalogs_.find(mountpoint);
-    if (iter != mounted_catalogs_.end()) {
-      if (iter->second != cache_hash) {
-        LogCvmfs(kLogCache, kLogDebug, "updating from %s to alien cache copy",
-                 iter->second.ToString().c_str());
-        success_code = catalog::kLoadNew;
+/**
+ * Loads (and fetches) a catalog by hash for a given mountpoint.
+ *
+ * Special case for root catalog: ctlg_info->root_ctlg_location must be given.
+ * If it is kCtlgLocationServer than another check for the most recent 
+ * manifest is done.
+ *
+ * @param [in, out] ctlg_info mandatory fields (input): mountpoint, hash
+ *         additional mandatory fields for root catalog: root_ctlg_location
+ *         output: sql_catalog_handle is set if catalog fetch successful
+ * @return kLoadUp2Date for root catalog that is already mounted
+ *         kLoadNew for any other successful load
+ *         kLoadFail on failure
+ */
+LoadReturn ClientCatalogManager::LoadCatalogByHash(CatalogInfo *ctlg_info) {
+  string catalog_descr = "file catalog at " + repo_name_ + ":" +
+    (ctlg_info->mountpoint.IsEmpty() ?
+      "/" : string(ctlg_info->mountpoint.GetChars(),
+                   ctlg_info->mountpoint.GetLength()));
+
+  catalog_descr += " (" + ctlg_info->hash.ToString() + ")";
+  string alt_root_catalog_path = "";
+
+  // root catalog needs special handling because of alt_root_catalog_path
+  CachedManifestEnsemble ensemble(fetcher_->cache_mgr(), this);
+  if (ctlg_info->mountpoint.IsEmpty()) {
+    if (fixed_alt_root_catalog_) {
+      alt_root_catalog_path = ctlg_info->hash.MakeAlternativePath();
+    }
+
+    // TODO(heretherebedragons) is this test necessary??
```

Comment:
I don't see that this is necessary, I think we should drop it.

Reply:
i just realized if we do this then we `CatalogInfo` needs to keep a copy of the manifest when `GetNewRootCatalogInfo` is being called